Add list users snippet, concept and testing (#732)
openfga/openfga.dev: docs/content/getting-started/perform-list-users.mdx

1. <SdkSetupPrerequisite />
2. You have [installed the SDK](./install-sdk.mdx).
3. You have [configured the _authorization model_](./configure-model.mdx) and [updated the _relationship tuples_](./update-tuples.mdx).
4. You have loaded `FGA_STORE_ID` and `FGA_API_URL` as environment variables.

</TabItem>

<TabItem value={SupportedLanguage.GO_SDK} label={languageLabelMap.get(SupportedLanguage.GO_SDK)}>

1. <SdkSetupPrerequisite />
2. You have [installed the SDK](./install-sdk.mdx).
3. You have [configured the _authorization model_](./configure-model.mdx) and [updated the _relationship tuples_](./update-tuples.mdx).
4. You have loaded `FGA_STORE_ID` and `FGA_API_URL` as environment variables.

</TabItem>

<TabItem value={SupportedLanguage.DOTNET_SDK} label={languageLabelMap.get(SupportedLanguage.DOTNET_SDK)}>

1. <SdkSetupPrerequisite />
2. You have [installed the SDK](./install-sdk.mdx).
3. You have [configured the _authorization model_](./configure-model.mdx) and [updated the _relationship tuples_](./update-tuples.mdx).
4. You have loaded `FGA_STORE_ID` and `FGA_API_URL` as environment variables.

</TabItem>

> <TabItem value={SupportedLanguage.PYTHON_SDK} label={languageLabelMap.get(SupportedLanguage.PYTHON_SDK)}>

Reviewer: Python is not yet supported

</TabItem>

<TabItem value={SupportedLanguage.CURL} label={languageLabelMap.get(SupportedLanguage.CURL)}>

1. <SdkSetupPrerequisite />
2. You have [configured the _authorization model_](./configure-model.mdx) and [updated the _relationship tuples_](./update-tuples.mdx).
3. You have loaded `FGA_STORE_ID` and `FGA_API_URL` as environment variables.

</TabItem>
</Tabs>

## Step By Step

Assume that you want to list all users of type `user` that have a `reader` relationship with `document:planning`:

### 01. Configure the <ProductName format={ProductNameFormat.ShortForm}/> API Client

Before calling the check API, you will need to configure the API client.

<Tabs groupId="languages">
<TabItem value={SupportedLanguage.JS_SDK} label={languageLabelMap.get(SupportedLanguage.JS_SDK)}>

<SdkSetupHeader lang={SupportedLanguage.JS_SDK} />

</TabItem>
<TabItem value={SupportedLanguage.GO_SDK} label={languageLabelMap.get(SupportedLanguage.GO_SDK)}>

<SdkSetupHeader lang={SupportedLanguage.GO_SDK} />

</TabItem>
<TabItem value={SupportedLanguage.DOTNET_SDK} label={languageLabelMap.get(SupportedLanguage.DOTNET_SDK)}>

<SdkSetupHeader lang={SupportedLanguage.DOTNET_SDK} />

</TabItem>

> <TabItem value={SupportedLanguage.PYTHON_SDK} label={languageLabelMap.get(SupportedLanguage.PYTHON_SDK)}>

Reviewer: No python

<SdkSetupHeader lang={SupportedLanguage.PYTHON_SDK} />

</TabItem>
<TabItem value={SupportedLanguage.JAVA_SDK} label={languageLabelMap.get(SupportedLanguage.JAVA_SDK)}>

<SdkSetupHeader lang={SupportedLanguage.JAVA_SDK} />

</TabItem>
<TabItem value={SupportedLanguage.CLI} label={languageLabelMap.get(SupportedLanguage.CLI)}>

<SdkSetupHeader lang={SupportedLanguage.CLI} />

</TabItem>
<TabItem value={SupportedLanguage.CURL} label={languageLabelMap.get(SupportedLanguage.CURL)}>

To obtain the [access token](https://auth0.com/docs/get-started/authentication-and-authorization-flow/call-your-api-using-the-client-credentials-flow):

<SdkSetupHeader lang={SupportedLanguage.CURL} />

</TabItem>
</Tabs>

### 02. Calling List Users API

> To return all users of type `users` that have have the `reader` relationship with `document:planning`:

Reviewer: ```suggestion
To return all users of type `user` that have have the `reader` relationship with `document:planning`:
```

<ListUsersRequestViewer
  authorizationModelId="01HVMMBCMGZNT3SED4Z17ECXCA"
  objectType="document"
  objectId="planning"
  relation="reader"
  userFilterType="user"
  expectedResults={{
    users: [{ object: { type: "user:anne" } }, { object: { type: "user:beth" } }],
    excluded_users: []
  }}
  skipSetup={true}
  allowedLanguages={[
    SupportedLanguage.JS_SDK,
    SupportedLanguage.GO_SDK,
    SupportedLanguage.DOTNET_SDK,

>     SupportedLanguage.PYTHON_SDK,

Reviewer: No python

4. You have loaded `FGA_STORE_ID` and `FGA_API_URL` as environment variables.

</TabItem>

<TabItem value={SupportedLanguage.JAVA_SDK} label={languageLabelMap.get(SupportedLanguage.JAVA_SDK)}>

1. <SdkSetupPrerequisite />
2. You have [installed the SDK](./install-sdk.mdx).
3. You have [configured the _authorization model_](./configure-model.mdx) and [updated the _relationship tuples_](./update-tuples.mdx).
4. You have loaded `FGA_STORE_ID` and `FGA_API_URL` as environment variables.

</TabItem>

<TabItem value={SupportedLanguage.CLI} label={languageLabelMap.get(SupportedLanguage.CLI)}>

1. <SdkSetupPrerequisite />
2. You have [configured the _authorization model_](./configure-model.mdx).
3. You have loaded `FGA_STORE_ID` and `FGA_API_URL` as environment variables.

</TabItem>

<TabItem value={SupportedLanguage.CURL} label={languageLabelMap.get(SupportedLanguage.CURL)}>

1. <SdkSetupPrerequisite />
2. You have [configured the _authorization model_](./configure-model.mdx) and [updated the _relationship tuples_](./update-tuples.mdx).
3. You have loaded `FGA_STORE_ID` and `FGA_API_URL` as environment variables.

</TabItem>
</Tabs>

## Step By Step

> Assume that you want to list all objects of type document that  user `anne` has `reader` relationship with:

Reviewer 2: ```suggestion
Assume that you want to list all users of type `user` that have a `reader` relationship with `document:planning`:
```

</TabItem>
<TabItem value={SupportedLanguage.CLI} label={languageLabelMap.get(SupportedLanguage.CLI)}>

<SdkSetupHeader lang={SupportedLanguage.CLI} />

</TabItem>
<TabItem value={SupportedLanguage.CURL} label={languageLabelMap.get(SupportedLanguage.CURL)}>

To obtain the [access token](https://auth0.com/docs/get-started/authentication-and-authorization-flow/call-your-api-using-the-client-credentials-flow):

<SdkSetupHeader lang={SupportedLanguage.CURL} />

</TabItem>
</Tabs>

### 02. Calling List Objects API

To return all documents that user `user:anne` has relationship `reader` with:

<ListUsersRequestViewer
  authorizationModelId="01HVMMBCMGZNT3SED4Z17ECXCA"
  objectType="document"
  objectId="planning"
  relation="reader"
  userFilterType="user"
  expectedResults={{
    users: [{ object: { type: "user:anne" } }],
    excluded_users: [{ object: { type: "user:beth" } }]
  }}
  skipSetup={true}
  allowedLanguages={[
    SupportedLanguage.JS_SDK,
    SupportedLanguage.GO_SDK,
    SupportedLanguage.DOTNET_SDK,
    SupportedLanguage.PYTHON_SDK,
    SupportedLanguage.JAVA_SDK,
    SupportedLanguage.CLI,
    SupportedLanguage.CURL,
  ]}
/>

> The result `document:otherdoc` and `document:planning` are the document objects that `user:anne` has `reader` relationship with.

Reviewer 2: Needs to be changed to align with ListUsers, not ListObjects.

:::caution Warning
The performance characteristics of the ListUsers endpoint vary drastically depending on the model complexity, number of tuples, and the relations it needs to evaluate. Relations with 'and' or 'but not' are more expensive to evaluate than relations with 'or'.

> :::

Reviewer 2: We have hunches about performance considerations but not enough to make assertions like this IMO. I would remove.

Author: Not a problem! Added in 0ba8391bc960ba565d1e0f7e5fa6ab1fae8460bd
